build_catalog.sh

```diff
@@ -1,105 +1,186 @@
-
-if [ $exists -eq 1 ]; then
-  log "Extracting previous version from bundle image"
-  docker create --name="tmp_$$" $BUNDLE_IMAGE:latest sh
-  tmp_dir=$(mktemp -d -t sa-XXXXXXXXXX)
-  pushd $tmp_dir
-    docker export tmp_$$ | tar -xf -
-    prev_version=`find . -name *.clusterserviceversion.* | xargs cat - | python3 -c 'import sys,yaml; print(yaml.safe_load(sys.stdin.read())["metadata"]["name"])'`
-    if [[ "$prev_version" == "" ]]; then
-      log "Unable to find previous bundle version"
-      exit 1
+[[ $count -gt 0 ]] && exit 1
+
+# We shouln't need set -u as we have checked for the vars before
+# but let's play safe
+set -euo pipefail
+
+# General vars to modify behaviour of the script
+DRY_RUN=${DRY_RUN:-true}
+DELETE_TEMP_DIR=${DELETE_TEMP_DIR:-true}
+REMOVE_UNDEPLOYED=${REMOVE_UNDEPLOYED:-false}
+
+temp_dir=$(mktemp -d)
+[[ "$DELETE_TEMP_DIR" == "true" ]] && trap 'rm -rf $temp_dir' EXIT
+
+engine_cmd="$CONTAINER_ENGINE --config=$CONFIG_DIR"
+
+# clone bundle repo containing current version
+saas_operator_dir_base="$temp_dir/saas-operator-dir"
+bundle_versions_file="$saas_operator_dir_base/$OPERATOR_NAME/${OPERATOR_NAME}-versions.txt"
+bundle_versions_package_file="$saas_operator_dir_base/$OPERATOR_NAME/${OPERATOR_NAME}.package.yaml"
+
+log "Cloning $BUNDLE_VERSIONS_REPO"
+if [[ -n "${APP_SRE_BOT_PUSH_USER:-}" && "${APP_SRE_BOT_PUSH_TOKEN:-}" ]]; then
+    bundle_versions_repo_url="https://${APP_SRE_BOT_PUSH_USER}:${APP_SRE_BOT_PUSH_TOKEN}@$BUNDLE_VERSIONS_REPO"
+else
+    bundle_versions_repo_url="https://$BUNDLE_VERSIONS_REPO"
+fi
+
+git clone --branch "$BRANCH_CHANNEL" "$bundle_versions_repo_url" "$saas_operator_dir_base"
+
+prev_operator_version=""
+removed_versions=""
+if [[ -f "$bundle_versions_file" ]]; then
+    log "$bundle_versions_file exists. We'll use to determine current version"
+    if [[ "$REMOVE_UNDEPLOYED" == "true" ]]; then
+        log "Checking if we have to remove any versions more recent than deployed hash"
+        # TODO: Move this out of here
+        deployed_hash=$(
+            curl -s "https://gitlab.cee.redhat.com/service/app-interface/-/raw/master/data/services/deployment-validation-operator/cicd/saas.yaml" | \
+                docker run --rm -i quay.io/app-sre/yq -r '.resourceTemplates[]|select(.name="deployment-validation-operator").targets[]|select(.namespace["$ref"]=="/openshift/app-sre-stage-01/namespaces/app-sre-dvo-per-cluster.yml")|.ref'
+        )
+
+        log "Current deployed hash is $deployed_hash"
+
+        new_bundle_versions_file=$(mktemp -p "$temp_dir")
+        delete=false
+        # Sort based on commit number
+        for version in $(sort -t . -k 3 -g "$bundle_versions_file"); do
+            if [[ "$delete" == false ]]; then
+                echo "$version" >> "$new_bundle_versions_file"
+
+                short_hash=$(echo "$version" | cut -d- -f2)
+
+                if [[ "$deployed_hash" == "${short_hash}"* ]]; then
```

Comment:
if `deployed_hash` is `master` it will never match any operator sha, so it will not delete anything... But it is a consequence of the implementation... maybe we shouldn't even try to delete if we get master.

Another angle is that we should never get `master` here as we're taking this from production...

Reply:
> I think this will make us miss the short-circuit on L101-4 when master corresponds to the currently-deployed version.

That should not affect that as that `prev_operator_version` comes from the `versions.txt` file, not from the value of `deployed_hash`

Reply:
> if `deployed_hash` is `master` it will never match any operator sha, so it will not delete anything... But it is a consequence of the implementation... maybe we shouldn't even try to delete if we get master.
> 
> Another angle is that we should never get `master` here as we're taking this from production...

I don't think we can make the assumption though can we?  Stage deployments will always be master or some other branch name.  We don't want to delete all newer revisions than what is on prod because that will prevent stage from testing newer versions.

Reply:
We can set `REMOVE_UNDEPLOYED` to `false` in staging.

Reply:
I'm not sure I understand that.  The operator, bundle, and catalog are built once independent of deployment environment.

Reply:
This is how it is done today. That's why there are two branches. In staging we keep everything, in prod we only keep what's been deployed

https://github.com/openshift/hive/blob/master/hack/app_sre_build_deploy.sh#L27-L31

Reply:
In this case I don't think we want to do this anymore.  Catalog images are built with the entire bundle upgrade train in it.  The only way to achieve this idea would be to edit the catalog and change the dependency tree to remove some elements and change the upgrade path.  This means the catalog that would get deployed to prod is not the same one tested in stage.  There are methods to allow skipping versions in the upgrade chain if that is needed.

Reply:
We need _something_ here to let us skip versions in prod. IIUC it's not just for terseness -- though that's a bonus. It's because sometimes versions are broken. And depending _how_ they are broken, accidentally deploying them in an upgrade chain could result in data loss, the need for manual recovery, etc.

Reply:
I'm saying this entire logic chain should be removed.  No pruning, no keeping separate list for prod and stage.  We have 1 list of versions that represents the upgraded chain for an operator.  The fact this data is kept in the catalog image pretty much requires it.

Reply:
I don't feel this is the place to review if the current strategy of separated catalogs per environment is the best we can get. While I understand that the solution is not perfect as we could have untested upgrade paths in the prod catalog, the point of @2uasimojo is very important: we need a way to remove broken versions in the prod catalog. I have implemented the two catalogs in 4cf9aec